### PR TITLE
WIP: Allocate param array entries dynamically.

### PR DESCRIPTION
**Pull request by paulidale**
This provides an alternative for the `OSSL_PARAM_construct_ `series of functions which allocate storage for the values rather than relying on local on stack storage.  The new `OSSL_PARAM_allocate_` and the existing `OSSL_PARAM_construct_` calls can be freely intermingled.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
This is intended to provide a way for OSSL_PARAM structures to escape a stack frame more easily.

Tests, documentation, string types, BN and bug fixing are still required.

**Comment by levitte** [comment]
I prefer this over #9246.
For clarity, documentation should contain examples on how to keep track of allocated OSSL_PARAM elements, such as:

``` C
    params[foo_pos = n++] = OSSL_PARAM_allocate_int("foo", 42);
```

... or:

``` C
    foo_pos = n;
    params[n++] = OSSL_PARAM_allocate_int("foo", 42);
```

**Comment by paulidale** [comment]
My intention is to have a flag bit (or type bit) to indicate allocated.  Because the function must return an OSSL_PARAM structure, error detection needs to occur later -- I've included a function for this here.  Likewise, a free function would be desirable.

**Comment by levitte** [comment]
What you're implying here as well as in #9139 is that the creator of the OSSL_PARAM array looses control of that array, i.e stops being the owner. I think that's a bad direction to follow, and adds a ton of complexity.

**Comment by mattcaswell** [comment]
I have a different approach to this in the works. At the moment it is EVP specific (because I wrote it for a very specific problem) but it could easily be made generic. It's not yet ready for prime time so I've not made a PR from it yet. See:

https://github.com/mattcaswell/openssl/blob/c3298b53dbe5c1543695549e28fe97a7fc7e2ea3/crypto/evp/evp_lib.c#L780-L819

The idea is to create a temporary "template" which then gets converted to an OSSL_PARAM at the end in a single OPENSSL_zalloc call (or OPENSSL_secure_zalloc as appropriate). The whole OSSL_PARAM structure can then be passed around as required, and the whole thing can be freed with a single OPENSSL_free call at the end.

**Comment by paulidale** [comment]
The creator doesn't lose control in either.

Here, the creator has complete control -- populate the OSSL_PARAM array, pass it to the consumer, free it.

In #9246 the creator needs to keep control, the creator allocates the memory and must free it.  Call the creator to get the param array, populate it, call the creator to use it and either call again to free it or have the use free as well.

**Comment by paulidale** [comment]
@mattcaswell, that's kind of heading in a similar direction to #9246 with its single allocation/free. I used a varargs function rather than building the description up piecemeal but the intent seems similar.

**Comment by mattcaswell** [comment]
Note that whatever solution is eventually chosen, it will need to be able to handle secure malloc so that we can pass private data as an OSSL_PARAM.

**Comment by richsalz** [comment]
The template approach mentioned in https://github.com/openssl/openssl/pull/9254#issuecomment-506264313 is what I did for #8377; defining the parameters was a separate step from allocating storage.

File: crypto/params.c [hunk 1]
```diff
@@ -33,12 +39,37 @@ static OSSL_PARAM ossl_param_construct(const char *key, unsigned int data_type,
...
+static OSSL_PARAM ossl_param_allocate(const char *key, unsigned int data_type,
+                                      void *data, size_t data_size)
+{
+    OSSL_PARAM res;
+    int flag = OWNER_FLAG_allocated;
+    void *p = OPENSSL_malloc(data_size);
+
+    if (p != NULL) {
+        if (data == NULL)
+            memset(p, 0, data_size);
```
Comment (richsalz) [inline crypto/params.c]: use ?: instead of the nested if.

File: crypto/params.c [hunk 2]
```diff
@@ -632,6 +716,49 @@ OSSL_PARAM OSSL_PARAM_construct_BN(const char *key, unsigned char *buf,
...
+    OSSL_PARAM res;
+    int flag = OWNER_FLAG_allocated;
+    int secure = 0;
+    void *p;
+
+    if (bn != NULL) {
+        if (bsize == 0)
+            bsize = (size_t)BN_num_bytes(bn);
+
+        if (BN_get_flags(bn, BN_FLG_SECURE)) {
```
Comment (richsalz) [inline crypto/params.c]: Add ```== BN_FLG_SECURE```

Also, this function needs comments.

File: crypto/params.c [hunk 3]
```diff
@@ -632,6 +716,49 @@ OSSL_PARAM OSSL_PARAM_construct_BN(const char *key, unsigned char *buf,
...
+            secure = 1;
+            flag |= OWNER_FLAG_secure;
+        }
+    }
+    if (secure)
```
Comment (richsalz) [inline crypto/params.c]: I prefer the ```?:``` operator, IMO

**Pull request closed without merging**